---
title: "Genomics Lab Session"
author: 'Kiley Hooker (PID: A15441609)'
date: "2/17/2022"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Section 1. Proportion of G/G in a population

Downloaded a CSV file from Ensemble < https://uswest.ensembl.org/Homo_sapiens/Variation/Sample?db=core;r=17:39894595-39895595;v=rs8067378;vdb=variation;vf=105535077#373531_tablePanel > 

Here we read this CSV file

```{r}
mxl <- read.csv("373531-SampleGenotypes-Homo_sapiens_Variation_Sample_rs8067378.csv")
head(mxl)
```

```{r}
table(mxl$Genotype..forward.strand.)
```

```{r}
table(mxl$Genotype..forward.strand.)/nrow(mxl) *100
```

## Section 4: Population Scale Analysis
One sample is obviously not enough to know what is happening in a population. You are
interested in assessing genetic differences on a population scale. So, you processed about ~230 samples and did the normalization on a genome level. Now, you want to find whether there is any association of the 4 asthma-associated SNPs (rs8067378...) on ORMDL3 expression.

> Q13: Read this file into R and determine the sample size for each genotype and their corresponding median expression levels for each of these genotypes.

How many samples do we have?

```{r}
expr <- read.table("rs8067378_ENSG00000172057.6.txt")
head(expr)
```

```{r}
nrow(expr)
```

```{r}
table(expr$geno)
```

For genotype A/A
```{r}
mean(expr[expr$geno == "A/A", "exp"])
```

For genotype A/G
```{r}
mean(expr[expr$geno == "A/G", "exp"])
```

For genotype G/G
```{r}
mean(expr[expr$geno == "G/G", "exp"])
```


> Q14: Generate a boxplot with a box per genotype, what could you infer from the relative expression value between A/A and G/G displayed in this plot? Does the SNP effect the expression of ORMDL3? 

```{r}
library(ggplot2)
```

Lets make a boxplot
```{r}
ggplot(expr, aes(geno, exp, fill=geno)) +
  geom_boxplot(notch=TRUE)
```

Question 14 answered in other PDF.


