---
title: "Transcriptomics and the analysis of RNA-Seq data"
author: 'Kiley Hooker (PID: A15441609)'
date: "2/22/2022"
output:
  pdf_document: default
  html_document: default
---

The data for this hands-on session comes from a published RNA-seq experiment where airway smooth muscle cells were treated with dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects (Himes et al. 2014).

## Read the countdata and coldata

We need two things the countdata and coldata (metadata about how and experiment was setup).
```{r}
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <-  read.csv("airway_metadata.csv")
```

Have a little look see:

```{r}
head(counts)
```

```{r}
head(metadata)
```
> Q1. How many genes are in this dataset? 38694

> Q2. How many ‘control’ cell lines do we have? 4

I always need to double check that the columns of my countdown coldata (metadata) match.

```{r}
metadata$id
```

```{r}
colnames(counts)
```

```{r}
metadata$id == colnames(counts)
```

```{r}
all(c(T, T, T, F))
```

I can use the `all()` function to make sure all my values match (i.e. all values are TRUE)

```{r}
all(metadata$id == colnames(counts))
```

## 2 Extract control and treated counts for comparison

First lets extract the control counts columns

```{r}
control.ids <- metadata[metadata$dex == "control", ]$id
control.counts <- counts[,control.ids]
#head(control.counts)
```

```{r}
# Take the mean count value per gene (i.e. row)
control.mean <- rowMeans(control.counts)
head(control.mean)
```

Now do the same thing for "treated" samples.

```{r}
treated.ids <- metadata[metadata$dex == "treated", ]$id
treated.counts <- counts[,treated.ids]
treated.mean <- rowMeans(treated.counts)
```

Now we can make a plot comparing treated vs control

```{r}
plot(treated.mean, control.mean,
     xlab= "log of drug treated counts",
     ylab= "log of control counts")
```

> Q5 (b).You could also use the ggplot2 package to make this figure producing the plot below. What geom_?() function would you use for this plot? geom_point

> Q6. Try plotting both axes on a log scale. What is the argument to plot() that allows you to do this? log = "xy"

```{r}
plot(treated.mean, control.mean, log = "xy",
     xlab= "log of drug treated counts",
     ylab= "log of control counts")
```

We are after changes in gene expression: treated vs control and this would represent points (i.e. genes) that do not like on the diagnal.

We like to work with log 2 values
```{r}
log2(20/20)
```

```{r}
# doubling
log2(40/20)
```

```{r}
log2(10/20)
```

```{r}
# quadrupling
log2(80/20)
```
Now lets calculate the log2 fold change

```{r}
log2fc <- log2(treated.mean/control.mean)
```

Store my work so far
```{r}
meancounts <- data.frame(control.mean, treated.mean, log2fc)
head(meancounts)
```

```{r}
sum(meancounts$log2fc > 2, na.rm=TRUE)
```

Filter our data to remove genes with zero expression values.

```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```
How many genes do we have left?

```{r}
nrow(mycounts)
```

> Q7. What is the purpose of the arr.ind argument in the which() function call above? Why would we then take the first column of the output and need to call the unique() function? It shows the row and column position where values of 0 are true.


A common threshold used for calling something differentially expressed is a log2(FoldChange) of greater than 2 or less than -2. Let’s filter the dataset both ways to see how many genes are up or down-regulated.

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
table(up.ind)
table(down.ind)
```

> Q8. Using the up.ind vector above can you determine how many up regulated genes we have at the greater than 2 fc level? 250

> Q9. Using the down.ind vector above can you determine how many down regulated genes we have at the greater than 2 fc level? 367

> Q10. Do you trust these results? Why or why not? No, not enough analysis has been done to determine if the results are significant. 


We are missing the stats! Are these differences significant?


# DESeq2 analysis

Let’s do this the right way. DESeq2 is an R package specifically for analyzing count-based NGS data like RNA-seq. It is available from Bioconductor. 

```{r}
library(DESeq2)
```

This package wants input in a specific way:

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex)
dds
```

Now we can run the DESeq2 analysis

```{r}
dds <- DESeq(dds)
```

To get the results back in a useful way we can use the `results()` function
```{r}
res <- results(dds)
res
```

## Volcano plots
Let’s make a commonly produced visualization from this data, namely a so-called Volcano plot. These summary figures are frequently used to highlight the proportion of genes that are both significantly regulated and display a high fold change.

```{r}
plot(res$log2FoldChange, res$padj)
```

```{r}
plot(res$log2FoldChange, -log(res$padj))
abline(h=-log(0.05), col="gray")
abline(v=c(-2,2), col="gray")
```

I want to polish this main results figure by adding color to the genes I will focus on next day. 

```{r}
# I will start by making a gray vector for everything
mycols <- rep("gray", nrow(res))

# Now I will overwrite the small padj values
mycols[res$padj < 0.005] <- "red"

# Now if my log2foldchange is small I will make them gray
mycols[ abs(res$log2FoldChange) < 2 ]  <- "gray"

plot(res$log2FoldChange, -log(res$padj), col=mycols,
     ylab="-Log(P-value)", xlab="Log2(FoldChange)")
abline(h=-log(0.05))
abline(v=c(-2,2))
```


# Adding annotation data

To help interpert our results we need to understand what the differentially expressed genes are. A first step here is to get the gene names (i.e. gene SYMBOLs).

For this I will install: 
- BiocManager::install("AnnotationDbi")
- BiocManager::install("org.Hs.eg.db")
```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```

What DB identifiers can I look up?
```{r}
columns(org.Hs.eg.db)
```

We will use the `mapIds()` function to translate between different ids.
```{r}
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",   # The format of our genenames
                     column="SYMBOL",     # The new format we want to add
                     multiVals="first")
head(res)
```

> Q11. Run the mapIds() function two more times to add the Entrez ID and UniProt accession and GENENAME as new columns called res$entrez, res$uniprot and res$genename.

```{r}
res$entrez <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     keytype="ENSEMBL",
                     column="ENTREZID",
                     multiVals="first")

res$uniprot <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     keytype="ENSEMBL",
                     column="UNIPROT",
                     multiVals="first")

res$genename <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     keytype="ENSEMBL",
                     column="GENENAME",
                     multiVals="first")

head(res)
```

# Pathway Analysis with R and Bioconductor

Here we use the GAGE package (which stands for Generally Applicable Gene set Enrichment), to do KEGG pathway enrichment analysis on our RNA-seq based differential expression results.

I need to install the gage package along with the pathview package for generating pathway figures from my results.

BiocManager::install( c("pathview", "gage", "gageData") )

Now load up the packages and have a peak at the first two pathways in KEGG

```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)

# Examine the first 2 pathways in this kegg set for humans
head(kegg.sets.hs, 2)
```

Recall that vectors can have a names attribute that helps with bookkeeping just like colnames and rownames.
```{r}
x <- c(40,70,20)
names(x) <- c("lisa", "xinqiu", "barry")
x
```

We need a vector of fold-change labeled with the names of our genes in ENTREZ format.

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
```

Now we can run the GAGE analysis passing in our foldchange vector and the KEGG genesets we are interested in.

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

Let's have a look at what is contained in this `keggres` results object (i.e. it's attributes).

```{r}
attributes(keggres)
```

```{r}
# Look at the first three down (less) pathways
head(keggres$less, 3)
```

Now I can map my foldchange results onto any KEGG pathway. I will do this manually first by selecting one of the pathway IDs from above. 

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05310")
```

![](hsa05310.pathview.png)

## Final step save our results.

```{r}
write.csv(res, file="deseq_results.csv")
```





